src/baldr/shortcut_recovery.h

```diff
@@ -0,0 +1,231 @@
+      if (shortcut->shortcut() & de.superseded()) {
+        edges.push_back(tile->header()->graphid());
+        edges.back().set_id(&de - tile->directededge(0));
+        break;
+      }
+    }
+
+    // bail if we couldnt find it
+    if (edges.empty()) {
+      LOG_TRACE("Unable to recover shortcut for edgeid " + std::to_string(shortcut_id) +
+                " | no superseded edge");
+      return {shortcut_id};
+    }
+
+    // seed the edge walking with the first edge
+    const DirectedEdge* current_edge = tile->directededge(edges.back());
+    uint32_t accumulated_length = current_edge->length();
+
+    // walk edges until we find the same ending node as the shortcut
+    while (current_edge->endnode() != shortcut->endnode()) {
+      // get the node at the end of the last edge we added
+      const NodeInfo* node = reader.GetEndNode(current_edge, tile);
+      if (!node)
+        return {shortcut_id};
+      auto node_index = node - tile->node(0);
+
+      // check the edges leaving this node to see if we can find the one that is part of the shortcut
+      current_edge = nullptr;
+      for (const DirectedEdge& edge : tile->GetDirectedEdges(node_index)) {
+        // are they the same enough that its part of the shortcut
+        // NOTE: this fails in about .05% of cases where there are two candidates and its not clear
+        // which edge is the right one. looking at shortcut builder its not obvious how this is
+        // possible as it seems to terminate a shortcut if more than one edge pair can be
+        // contracted... NOTE: because we change the speed of the edge in graph enhancer we cant use
+        // speed as a reliable determining factor
+        if (begin_node != edge.endnode() && !edge.is_shortcut() &&
+            edge.forwardaccess() == shortcut->forwardaccess() &&
+            edge.reverseaccess() == shortcut->reverseaccess() && edge.sign() == shortcut->sign() &&
+            edge.use() == shortcut->use() && edge.classification() == shortcut->classification() &&
+            edge.roundabout() == shortcut->roundabout() && edge.link() == shortcut->link() &&
+            edge.toll() == shortcut->toll() && edge.destonly() == shortcut->destonly() &&
+            edge.unpaved() == shortcut->unpaved() && edge.surface() == shortcut->surface()/* &&
+          edge.speed() == shortcut->speed()*/) {
+          // we are going to keep this edge
+          edges.emplace_back(tile->header()->graphid());
+          edges.back().set_id(&edge - tile->directededge(0));
+          // and keep expanding from the end of it
+          current_edge = &edge;
+          begin_node = tile->header()->graphid();
+          begin_node.set_id(node_index);
+          accumulated_length += edge.length();
+          break;
+        }
+      }
+
+      // if we didnt add an edge or we went over the length we failed
+      if (current_edge == nullptr || accumulated_length > shortcut->length()) {
```

Comment:
The reason it's not an "approximate equality" is because not all parts of the shortcut is required to be recovered?

Reply:
Oh, the test for shorter length is below. I see.